### DISCUSSION TITLE
Ideas and Features

### DISCUSSION THREAD

**Discussion by TheCutter** [TheCutter/RoboMow2MQTT#3] [Ideas]
- Status update (Current State, percentage done, time, warning, errors, ...)
- Trigger Actions (Start Mowing, stop mowing, go home, ...)
- Connection via WiFi to home network and BLE to Robomow

**Comment by geelenbert** [comment]
I would be using this with Home assistant, but MQTT would work great!

The main goal should be to focus on just wifi, not LoRa or other long range communications. It would probably be easer to create good wifi coverage over your lawn than other options.

I would agree with the features above. Battery percentage would be very nice (for monitoring battery life and condition).

Maybee a nice to have would be a GPS location ?

**Comment by TheCutter** [comment]
Yes, the ESP32 will connect via WiFi.
What do you mean by GSP location? My Mower doesn't have a GPS Sensor installed.

**Comment by promy** [comment]
The original (arjanmels) was using TTGO T-Beam board  . This board has a gps sensor on it.
The above features would cover all my needs ;-)

**Comment by geelenbert** [comment]
Although it would be fun to see where the mower has been, it is totally unnecessary for basic operations. 

I think zone support would be the next thing to add.

**Comment by TheCutter** [comment]
Okay got it now. My plan is to put the ESP Board in a fixed place next to my garden electrics, so GPS does not make any sense to me.

**Comment by promy** [comment]
> Okay got it now. My plan is to put the ESP Board in a fixed place next to my garden electrics, so GPS does not make any sense to me.

Not so sure thats the best idea since the BLE distance is limited (my garden is to big for full coverage). But thats every one's choise. I was thinking of adding  a battery and have it charged every time the mower charges. (or steal some power from the mower)

**Comment by TheCutter** [comment]
My plan is to get that thing running for the next season here in Germany. Which will be around April 2022.

**Comment by dagope** [comment]
Are you planning use ESPHome with BLE for easy integration in Home Assistant?
I think the principal features to control remote our Robomow are: start, stop and go home.

**Comment by TheCutter** [comment]
I don't plan to use ESPHome, just plain MQTT connection which will be easily integratable with all plattforms.

**Comment by promy** [comment]
Since we are april 2022 :-) would there be any progress? thanks!

**Comment by TheCutter** [comment]
I'm sorry, but I didn't have much time to work on this project. I got some problems fixed, but still no version I could release.
Will keep you updated!

**Comment by arcidodo** [comment]
any news on this? :-)

**Comment by TheCutter** [comment]
I'm so sorry guys to keep you waiting, but since I became a father 6 weeks ago, time is very limited to work on this project.
But everytime my Robomow starts mowing in the rain I really want this project to happen...
Maybe I can publish a very limited version in the next weeks or months and someone else can add some more stuff.

**Comment by arcidodo** [comment]
First of all, congratulations!🍼

If you upload your version to GitHub maybe we can continue to develop this. 

I already did an hardware mod so I can start/stop/gohome. But I can't get the Status of the mower. So this would be very helpful for me. 

Hwmod:
![image](https://user-images.githubusercontent.com/678514/172946355-bdc2df1f-ac3b-4089-8e1b-8b8ec0918db3.jpeg)

Ha:
![image](https://user-images.githubusercontent.com/678514/172946501-5d9e67f1-0037-46b5-8844-8ef2935ec2a6.png)

**Comment by TheCutter** [comment]
Nice! I heard about it. It's simulating button presses to control the mower, right?
Let me get the basic stuff running and I will publish it.

**Comment by mattncsu** [comment]
I haven't implemented it but I think you could check the status by watching the power draw on the plug the base station is plugged in to. My RS630 draws about 19W while cutting, 100W fast charging and around 25W while sitting idle on the charger. If you get a smart plug with current monitoring, you could poll it to get an idea of what the mower is doing. I want to implement a watchdog to notify me if the mower is stuck by watching for the current draw to be under 20W for two hours for example. Just haven't made the time to script it yet.

**Comment by arcidodo** [comment]
> Nice! I heard about it. It's simulating button presses to control the mower, right? Let me get the basic stuff running and I will publish it.

yea its simulating button press. and i read the "error" and the "child lock" led to get some status. :-) it just works but its really "fire and forgot" so no feedback.

**Comment by mattncsu** [comment]
A few of the Robomows I've seen have a serial port. It looks like the rx50 does per this parts diagram- maybe you can communicate over serial

**Comment by TheCutter** [comment]
> A few of the Robomows I've seen have a serial port. It looks like the rx50 does per this parts diagram- maybe you can communicate over serial

As far as I remember, the serial port is only active while charging/standing still. There are some limitations to that.

**Comment by bergh2** [comment]
Very interesting project!!
Just curious, have any of you tried tinkering with the power box? Should be a way of reading mower status (mowing/charging), simulate button-press to prevent the that the mower starts a new sessions and also read status of perimeter wire…
Any thoughts?

**Comment by TheCutter** [comment]
Guys, a quick update: I finally got the BLE communication working and just received the first answer from my Robomow!
Now it should be just coding all the required messages. But it will still take some time...

**Comment by arcidodo** [comment]
> Guys, a quick update: I finally got the BLE communication working and just received the first answer from my Robomow!
> 
> Now it should be just coding all the required messages. But it will still take some time...

Nice! If you want to share your code then I can try to help with coding😁

**Comment by geelenbert** [comment]
I would like to know if we can help with the coding.

Is there any new code available we can try our self?

**Comment by promy** [comment]
@TheCutter  any updates on the code. Would be nice to implement it here for this season ;-)

**Comment by geelenbert** [comment]
@TheCutter please share your code so we can continue working on it this season

**Comment by bergh2** [comment]
I would also be happy to contribute to this project :)

**Comment by TheCutter** [comment]
I'm so sorry guys. I really didn't have the time to work on the project. I will finish the code to send and receive one message from the robomow and after a code cleanup I will upload it for you to work on.

**Comment by TheCutter** [comment]
Just wanted to let you know that I switched to PlatformIO with the project and got it running again today. Will keep you informed when the project is ready for development.

**Comment by Nickman87** [comment]
Hi @TheCutter, would love to jump in on this project to get some control over our RoboMow. Any indications when you feel comfortable getting the project on the GitHub? :)

**Comment by TheCutter** [comment]
Hey @Nickman87 that sounds great. I just want to confirm that everything is working and add some documentation. Just give me a few more weeks. Free time is still very limited at the moment. But I don't want to push something that is not working.

**Comment by TheCutter** [comment]
I finally found some time to work on the project, but I constantly get "Brownout Detector triggered" exceptions with my ESP32...
Haven't found a solution yet. Which is really frustrating. Will keep you guys updated.

**Comment by mattncsu** [comment]
You need a better power supply for the ESP32. Brownout is triggered when voltage gets too low, [this](https://github.com/espressif/arduino-esp32/issues/2580) makes me think that value is 2.4V

**Comment by geelenbert** [comment]
Can you maybee describe your setup?
What board, what mower, how did you connect the power?

**Comment by TheCutter** [comment]
> You need a better power supply for the ESP32. Brownout is triggered when voltage gets too low, [this](https://github.com/espressif/arduino-esp32/issues/2580) makes me think that value is 2.4V

It always worked with this configuration, but after hooking everything up again a few days ago, I constantly get the Brownout message. Don't know why... 

> Can you maybee describe your setup? What board, what mower, how did you connect the power?

I got an ESP32 NodeMCU Dev Kit Board connected to an ESP-Prog to get proper debugging.
The Board is getting the power through the ESP-Prog. As I said it always worked, but stopped working now.
I also tried different versions of the Espressif SDK, no luck. Only disabling the Brownout Detection helps, but that can not be the real solution...

**Comment by mattncsu** [comment]
If you have any physical connections (serial port, buttons, etc) to your robot, try adding series resistors to limit the current.  You could also check all your joints to make sure there are no shorts, stray wires, loose wires, etc. You could add a capacitor on the power input to give the board a little boost when the wifi/ble transmits.  Lastly, if you're convinced there isn't an actual brownout going on, you can disable brownout detection: https://iotespresso.com/how-to-disable-brownout-detector-in-esp32-in-arduino/

**Comment by TheCutter** [comment]
The Brownout Error is gone now, with a separate power supply. But now I don't get the WiFi/MQTT Stuff running, because of an unhandled exception. It really sucks, because I spent 2 days now to try to get it running again and it still doesn't work...
I will just cleanup the code and upload it to GitHub for you guys to continue, as I don't really have the time for this anymore.

**Comment by MortenVinding** [comment]
> I will just cleanup the code and upload it to GitHub for you guys to continue, as I don't really have the time for this anymore.

sounds really interesting!

I have looked at the other code, but it will not compile because of a discontinued library, or something.

I tried to remove some of the code there is not strictly necessary, but it’s way to complex for my coding skills.

**Comment by TheCutter** [comment]
I've got some good news. I found the problem with the WiFi/Mqtt and everything is working again. 
I will also check the BLE Connection and maybe implement some methods as examples for future BLE communication messages.

**Comment by MortenVinding** [comment]
> I've got some good news.

That sound awesome!
I was just about to start digging in to the code again to see if I could isolate the bluetooth stuff, but now I think I will wait and see what you have.

Would be really nice to have it ready for the season 👍

**Comment by TheCutter** [comment]
Some more good news. I got the whole project up und running with PlatformIO including debugging with ESP-Prog.

**Comment by arcidodo** [comment]
awesome news! :-)

**Comment by MortenVinding** [comment]
> Some more good news. I got the whole project up und running with PlatformIO including debugging with ESP-Prog.

Any news?
I think we are a few eagerly waiting for this :)
Sorry for being so impatient :)

**Comment by TheCutter** [comment]
I know, I know. But I would like to give out a version thats working.
The Config can already be read (including the Robomow family (RS, RC, RX), mainboard version, etc.)
But I would like to get the Bluetooth message running to start a mowing operation.

I'm already feeling bad for lettings you guys wait so long, but just give me a little bit more time. Thx!

**Comment by MortenVinding** [comment]
> I'm already feeling bad for lettings you guys wait so long, but just give me a little bit more time.

Don’t feel bad.
We are just like little kids waiting for x-mas 😆

**Comment by Moeso69** [comment]
I have a RK model could I help with something ?!

**Comment by dannybeeckman** [comment]
I´ve got a RT300 and a BLE sniffer. I can help with those if you want.

**Comment by geelenbert** [comment]
What is the status of the initial commit? I’m anxiously waiting to get started with the project on my own mower

**Comment by TheCutter** [comment]
Oh okay, so you guys already got the new models. I have an old RS Series Mower and only decompiled the app version 2 for the RS, RC and RX mowers.

What would be useful for the new RT and RK models are the BLE Characteristics GUIDs. For the RS it is:
Service: ff00a501-d020-913c-1234-56d97200a6a6 with three characteristics like:
ff00a502-d020-913c-1234-56d97200a6a6, ff00a503-d020-913c-1234-56d97200a6a6.

And I need the BLE Messages being send for getting the config (model, version, ...) of the mower + the message for starting the mowing operation.

For me it was very helpful to get the application log for the Robomow App via ADB. The BLE Sniffer was not helpful at all.
Even better is decompiling the original Robomow App and have a look in the code and the messages constructed.

If you could provide a application log with some informations at which timestamp in the log you activated which function of the mower what be useful.

**Comment by TheCutter** [comment]
> What is the status of the initial commit? I’m anxiously waiting to get started with the project on my own mower

Still working on the first version. I moved my mower to the back of the garden for this season and the ESP32 I used before wasn't able to get a stable communication with the mower. So I bought some new ones with external antennas.
I tried to get them working a few days ago, but as always there were some other problems regarding WiFi connectivity.

If I get that working and were able to implement the basic functionality for starting a mowing operation via MQTT I will check in the source code. That's the bare minimum I would like to get working on my side.

**Comment by TheCutter** [comment]
The new ESP32 board with external antenna works like a charm and I got very stable and fast connection to the Robomow.
Next step: Get State Information and start a mowing operation.

**Comment by geelenbert** [comment]
so is there any code to post ?

**Comment by TheCutter** [comment]
I've got the state information, but could not start a mowing operation yet. 
Are you an experienced c++ developer @geelenbert ?

**Comment by geelenbert** [comment]
no i am not.

**Comment by TheCutter** [comment]
Ah okay. I thought so. 
I fixed a nasty bug today, so I can reconnect over and over again now, but still need to find a memory leak. Next: parsing state message and trying to send start mowing command. 

Did you guys got the ble messages for the RT and RK?

**Comment by geelenbert** [comment]
But i do know some code. 
There are maybee others here that are better at coding. 

When you post your initial commit, we cam try it ourselfs and help if we can

**Comment by TheCutter** [comment]
Okay. When I push the initial commit it will only work with an robomow RS. Maybe RC and RX. Because I only had a look at the old version of the robomow app for these three mowers.

**Comment by arcidodo** [comment]
Let’s do it! Then we can try it ourself and get on with the code 😊 I am not a expert in c++ but I can read/write some code.

**Comment by pergolafabio** [comment]
Hey, found this repo, but the code is gone? It's an empty repo?

Thnx, :-)

**Comment by TheCutter** [comment]
Hey. It is not gone, it is not released yet. ;)

**Comment by pergolafabio** [comment]
> Hey. It is not gone, it is not released yet. ;)

Ah ok , makes sense then... I have an RC306 model, will it also work? Or do I need to provide some debug?

When will you upload the code?

**Comment by pergolafabio** [comment]
Hi, do you have an update for us?

thnx in advance

**Comment by TheCutter** [comment]
Unfortunately not... 
As I said earlier RS/RC/RX models will probably work, because of the App Version 2.
For the newer models I need some informations from the owners of these models.

**Comment by pergolafabio** [comment]
I still have the old version, rc306...

Ok, I'll wait.. any eta? :-)

**Comment by pergolafabio** [comment]
Hey @TheCutter , are you still working on this?

Thnx for update

**Comment by TheCutter** [comment]
Hi, I didn't have any time to work on this project in the last months...
There were other things way more important than this. I'm sorry to let 
you guys wait so long.

**Comment by pergolafabio** [comment]
no problem, the robomow is now in Winter sleep anyway :-)

**Comment by pergolafabio** [comment]
Hi, mowing season starts soon, any update on your code?

thnx

**Comment by TheCutter** [comment]
Not really. But I will publish the code, even if it is in an unfinished state. Maybe someone else can help with the development.
The Problem is my Robomow is in winter sleep too and I can't work with it and test in this state.

**Comment by pergolafabio** [comment]
Yeah, indeed, a dev branch would be nice
I know a  c+ porgrammer, who also got a Robomow, i can indeed share the code
I know also python, so i can have a look too later

**Comment by pergolafabio** [comment]
I will also post it to Home Assistant / Openhab forums, there are probably some gurus there too

**Comment by pergolafabio** [comment]
can you keep us posted when you create an Dev branch ?

**Comment by pergolafabio** [comment]
hey @TheCutter , were you able to upload the code? Mower season is almost starting :-)

**Comment by TheCutter** [comment]
The Code is online now. Have fun.
The Connection and getting the information about the robomow family works, but thats it.

**Comment by pergolafabio** [comment]
That's great, will have a look asap, need to reconnect my ESP first outside

**Comment by pergolafabio** [comment]
> The Code is online now. Have fun. The Connection and getting the information about the robomow family works, but thats it.

Are you also on Home Assistant or Openhab? Am ik allowed to share this repo?

**Comment by TheCutter** [comment]
No I'm not there. Yes, you can share it. Any help is welcome.

**Comment by pergolafabio** [comment]
Ok, do you use anything else maybe for smart home?

**Comment by TheCutter** [comment]
Yes. Node-Red, MQTT, Homematic, Custom solutions with ESPs.

**Comment by MortenVinding** [comment]
I'm very much looking forward to get MQTT support for my Robomow :)

In the mean time I have played a bit around with:
https://www.skyynet.de/mowgli_connect.php

But it is all in german, so I made my own repo about it and also made a really dirty script to translate the german in the (binary) file.
https://github.com/MortenVinding/Robomow-ESP32

Unfortunately it seems he won't release the sourcecode since he is afraid that Friendly Robots (maker of Robomow) will complain about infringing there closed protocol :(

Let me know if it's useful and if I should change anything.

**Reply by TheCutter** [reply]
Hey Morten,

I know the software Mähspitzel and already talked to the dev. He send me the source code, but it's not really usable, because it's a mess (all in one very large file) and the way the communication is handled is very different from my approach.

I want to have classes that handle the specific serialization and deserialization of the messages. So it's east to extent.

My plan is to continue the development, but as stated above it's a timing and priority problem. :)

**Reply by geelenbert** [reply]
I'm going to try to interface directly with the RS232/service port of the RS mower.
This should eventually result in just a plug and play solution on the service terminal port.

From my initial search it looks like the serial protocol of the service port and bleutooth port are equal.
The robotool service tool kind of confirms this, as you can choose between direct serial connection or ble.

I've read  that you (@TheCutter )had the source code of the Mähspitzel? Would it be possible to post or send the defines fo the serial protocol? Or get any other info of this ?

If what i'm thinking is going to work, it should be quite easy implementation with just a serial interface to the robowmow

**Reply by TheCutter** [reply]
Using the serial port won't work, because it is only active while the robomow is not in action. I've read that somewhere. 
Mähspitzel is also using Bluetooth to connect to the mower.

**Reply by geelenbert** [reply]
Well i’m gonna play with it some more. From what i  can see from the service manuals etc… is that the GSM module connects to the same service port. 

So maybe it is a matter of telling the mower that there is a GSM module connected or something like that. 

Either way, if there is anything you can share about the serial protocol over bluetooth, it would probably be very useful when trying to decode data.

**Reply by Cerize110** [reply]
I've read that @TheCutter you had the source code of the Mähspitzel? Would it be possible to post or send it ? I would like to continue his work for the RK version without starting from scratch. I tried to contact him without success, it no longer seems active

**Comment by MortenVinding** [comment]
> … is that the GSM module connects to the same service port.

It isn't, at least not on my 2017 RC model.
Maybe on the newer models?

**Comment by arcidodo** [comment]
https://github.com/MortenVinding/Robomow-ESP32


i found another repo, didn't test it.

**Comment by pergolafabio** [comment]
Hey @TheCutter , I'm using the project from @MortenVinding , the only thing there missing is the manual mow... You can steer the mower, but the blade is not turning...
In the old Robomow app it's possible to manually mow, do you know the BT command to activate the blades?

**Comment by promy** [comment]
I tried to install the @MortenVinding on a TTFO T-Beam, but it seems that i cannot upload the bin file (i keep on getting "fehler"). Did anyone manage to use a different board than the LOLIN D32 PRO?

**Comment by pergolafabio** [comment]
I had the same issue, make sure to setup the Arduino tools with the correct settings

**Comment by pergolafabio** [comment]
See here :

https://github.com/MortenVinding/Robomow-ESP32/issues/10

You need offcourse an esp with enough space

**Reply by promy** [reply]
Yes, that was the problem ;-) Thanks!! 
I got a esp-cam board running ;-) the T-Beam didn't have those settings....

**Reply by promy** [reply]
The T-Beam is also working ;-) but set as a Wrover in arduino , not as T-beam

**Comment by pergolafabio** [comment]
Hey @TheCutter , is it possible to share the code from Mähspitzel, I'm running that project now, works quite good, you can also control/move the mower, but there is no option to activate the blades when moving... I want to change that command..

Or can you compile a new bin file that the movement is also with the blades on? Seems the command to activate the blades is the same for RS and RC models, or maybe add a new toggle function?

**Reply by promy** [reply]
i would also be interested to change pushbullet to pushover ;-)

**Comment by promy** [comment]
Does anyone know where to buy the connector (male + female) to connect in between the battery and the robomow locally in Europe (belgium)? I can only find this T-style 2 pin in china?

**Reply by MortenVinding** [reply]
> Does anyone know where to buy the connector (male + female) to connect in between the battery and the robomow locally in Europe (belgium)? I can only find this T-style 2 pin in china?

You must tell for which robot, because my 2017 RC304u does only have one big multi connector.

I just build the whole thing in to the battery:
https://github.com/MortenVinding/Robomow-ESP32/tree/main/images

**Reply by promy** [reply]
Sorry, the RS630, it has a 10pin connector and a big 2pin T-style connector (with blades of 6.3mm)
This is my [battery](https://www.123accu.nl/image/Robomow_BAT6000A__MRK6103A_accu_25.6_V_6000_mAh_Li-Fe_123accu_huismerk_ARO00119_m1_big.jpg)

**Comment by TheCutter** [comment]
Sorry guys for the long delay. But as I said in #6 I got the first commands working. It is really early in development, but I can push the changes in the next days for you to try out.

**Reply by TheCutter** [reply]
> Sorry guys for the long delay. But as I said in #6 I got the first commands working. It is really early in development, but I can push the changes in the next days for you to try out.

I commited the latest changes which should allow the control of the Robomow RS mower via MQTT.

**Reply by pergolafabio** [reply]
Can you also write a guide how to get starting?

**Comment by pergolafabio** [comment]
Aha. Good news!!

What I miss in the Robomow app. What they removed a few years ago, was the option to manually mow, would be great to include this again...

Here is the info to send a manual mow...


For the RC the command string is
"aa 0a 1f 1a 10 00 64 00 00" +CKS
10 is a counter; the higher nibble is counted up for every time next command (-> 20, 30, 40...)
To activate the blade you have to set bit 2 of the lower nibble. -> (12, 22, 32...)
The next byte "00" is the direction (angle / 2) and 64 = (dec 100 %) is the speed.

**Reply by motogon** [reply]
> Aha. Good news!!
> 
> What I miss in the Robomow app. What they removed a few years ago, was the option to manually mow, would be great to include this again...
> 
> Here is the info to send a manual mow...
> 
> For the RC the command string is "aa 0a 1f 1a 10 00 64 00 00" +CKS 10 is a counter; the higher nibble is counted up for every time next command (-> 20, 30, 40...) To activate the blade you have to set bit 2 of the lower nibble. -> (12, 22, 32...) The next byte "00" is the direction (angle / 2) and 64 = (dec 100 %) is the speed.

On the RS625, I couldn't get the blades to start that way :(

**Reply by pergolafabio** [reply]
Maybe only for RC indeed, maybe sniff the old manmow app?

**Reply by motogon** [reply]
I don't have that app, and I couldn't find it :(

**Comment by TheCutter** [comment]
I know. It was removed due to safety regulations of moving blades controlled from distance. Some time ago there was an app called ManMow that allowed this operation. But I can't find it anymore, maybe it is somewhere in the Internet. ;)

**Comment by pergolafabio** [comment]
I know, I have the manmow :-)
Would be great to use the MQTT version for this... everything from within HA

**Comment by pergolafabio** [comment]
[20240806_ManMow.zip](https://github.com/user-attachments/files/28172763/20240806_ManMow.zip)

**Reply by motogon** [reply]
It won't download from the link :(

**Reply by motogon** [reply]
<img width="1298" height="165" alt="изображение" src="https://github.com/user-attachments/assets/e2f77ba1-bad4-421b-98b8-fcd415cb2fd8" />

Could you please send it to my email?

dend.mob at gmail.com
Thanks

**Reply by pergolafabio** [reply]
just uploaded it: https://we.tl/t-5oWvYcnVUWfMRpTw

**Reply by motogon** [reply]
Thanks, I downloaded it.
I'll give it a try.

**Comment by mattncsu** [comment]
I haven't had my Robomow in a few years but I remember getting the old
version from apkmirror to keep the manual cutting function

**Comment by pergolafabio** [comment]
That will not work anymore, if you install old version. It will ask to update once you start it...the manmow is modified to be not the real "Robomow" app

**Comment by mattncsu** [comment]
Bummer. I know the old version I was keeping was not Robomow branded. It
was another lawn brand made by Robomow (that I forget now) so I was able to
keep both side by side

**Reply by mattncsu** [reply]
It was the cub cadet brand.  Try the oldest version on apkpure... https://apkpure.com/cub-cadet-xr-2-0/com.cubcadet.app/versions  I would use the cub cadet app if i wanted to mow manually and keep the robomow app up to date in case there were any features added.  both apps coexisted fine on the phone.

**Reply by pergolafabio** [reply]
And that worked with Robomow RC/RS too ?

**Reply by mattncsu** [reply]
I used it with an RS630
